src/charm_relation_building_blocks/relation_handlers/application_data.py

```diff
@@ -0,0 +1,187 @@
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""Relation handler base classes for relations using application data."""
+
+from typing import List, Optional, Type, Union
+
+from ops import BoundEvent, CharmBase, CharmEvents, EventBase, EventSource, Object
+
+from charm_relation_building_blocks.data_models import DatabagModel, DataValidationError
+
+
+class DataChangedEvent(EventBase):
+    """Charm Event triggered when the relation data has changed."""
+
+
+class ReceiverCharmEvents(CharmEvents):
+    """Events raised by the data receiver side of the interface."""
+
+    data_changed = EventSource(DataChangedEvent)
+
+
+class Receiver(Object):
+    """Base class for the receiver side of a generic uni-directional application data relation."""
+
+    on = ReceiverCharmEvents()  # type: ignore[reportAssignmentType]
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str,
+        data_model: Type[DatabagModel],
+        refresh_event: Optional[Union[BoundEvent, List[BoundEvent]]] = None,
+    ) -> None:
+        """Initialize the Receiver object.
+
+        Args:
+            charm: The charm instance that the relation is attached to.
+            relation_name: The name of the relation.
+            data_model: The pydantic data model class to use for instantiating data instances.
+            refresh_event: An event or list of events that should trigger this library to process its relations.
+                           By default, this charm already observes the relation_changed event.
+        """
+        super().__init__(charm, relation_name)
+
+        self._charm = charm
+        self._relation_name = relation_name
+        self._schema = data_model
+
+        if not refresh_event:
+            refresh_event = []
+        if isinstance(refresh_event, BoundEvent):
+            refresh_event = [refresh_event]
+        for ev in refresh_event:
+            self.framework.observe(ev, self.on_relation_changed)
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_changed, self.on_relation_changed
+        )
+
+    def __len__(self):
+        """Return the number of related applications."""
+        return len(self.get_relations())
+
+    def on_relation_changed(self, _: EventBase) -> None:
+        """Handle when the remote application data changed."""
+        self.on.data_changed.emit()
```

Comment:
Continuing discussion from [here](https://github.com/canonical/istio-k8s-operator/pull/44#discussion_r1938031038), where it is raised that this event adds no value here but might be a pattern we should adopt anyway.